### PR TITLE
MMCA-4809 | Move the complex logic written in the view to view model

### PR DESCRIPTION
**Pull request by Neozxc**
**Acceptance Criteria:**
Ensure the complex logic in 'views/securities/security_statements.scala.html' is moved to the view model.

**Dev Tasks:**
- [x] Move complex logic from security_statements.scala -> SecurityStatementsViewModel.
- [x] Add unit tests for components and view model.

### CODE REVIEW

File: app/viewmodels/SecurityStatementsViewModel.scala [hunk 1]
```diff
@@ -16,9 +16,259 @@
...
+                                (implicit messages: Messages): HtmlFormat.Appendable = {
+    val statementContent = statementsForEori.currentStatements.zipWithIndex.flatMap {
+      case (statement, index) if isCsv != statement.hasPdf =>
+        Some(generateStatementRow(statement, historyIndex, index, isCsv))
+      case _ => None
+    }
+
+    dlComponent(
+      content = HtmlFormat.fill(statementContent),
+      classes = Some("govuk-summary-list statement-list"),
```
Comment (wg-hmrc) [inline app/viewmodels/SecurityStatementsViewModel.scala:134]: @Neozxc  The Gov.UK Design system already has the [Summary list component ](https://design-system.service.gov.uk/components/summary-list/) already built you could import this instead which could reduce the need for custom components like `dtComponent` and `ddComponent`, making the code easier to manage.

File: test/views/components/H1Spec.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        h1ComponentWithId.text() mustBe message(msgKey)
+
+        val h1: Elements = h1ComponentWithId.select("h1")
+        h1.first().attr("id") mustBe id
+      }
+
+      "classes and id have been provided along with message key" in new Setup {
+        h1ComponentWithClassAndId.text() mustBe message(msgKey)
+
+        val h1: Elements = h1ComponentWithClassAndId.select("h1")
```
Comment (tyagikumarvineet) [inline test/views/components/H1Spec.scala:61]: Better to give a line break after this line to differentiate between result and assertions

File: test/views/components/description_list/DescriptionListSpec.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import views.html.components.description_list.{dd, dl, dt}
+
+class DescriptionListSpec extends SpecBase {
+
+  "dd component" should {
+
+    "display correct contents" when {
+
+      "only content has been provided" in new Setup {
+        val dd: Elements = ddComponent.select("dd")
```
Comment (tyagikumarvineet) [inline test/views/components/description_list/DescriptionListSpec.scala:36]: Please add a line break after this line to differentiate between result and assertions. Same for other tests

File: app/viewmodels/SecurityStatementsViewModel.scala [hunk 4]
```diff
@@ -16,9 +16,259 @@
...
+        url = appConfig.requestedStatements(SecurityStatement))
+    } else {
+      HtmlFormat.empty
+    }
+  }
+
+  private def generateCurrentStatements(statementsForAllEoris: Seq[SecurityStatementsForEori])
+                                       (implicit messages: Messages): HtmlFormat.Appendable = {
+    if (hasCurrentStatements(statementsForAllEoris)) {
+      HtmlFormat.fill(statementsForAllEoris.zipWithIndex.flatMap {
```
Comment (tyagikumarvineet) [inline app/viewmodels/SecurityStatementsViewModel.scala:84]: Could you please move this functionality into a separate method to make is more readable like below? I chose a method name, check if that is meaningful for you but we need to keep a name on the same line to understand the flow easily

  private def generateCurrentStatements(statementsForAllEoris: Seq[SecurityStatementsForEori])
                                       (implicit messages: Messages): HtmlFormat.Appendable = {
    if (hasCurrentStatements(statementsForAllEoris)) {
      HtmlFormat.fill(populateCurrentPdfAndCsvStatements(statementsForAllEoris))
    } else {
      generateNoStatementsParagraph
    }
  }

  private def populateCurrentPdfAndCsvStatements(statementsForAllEoris: Seq[SecurityStatementsForEori])
                                                (implicit messages: Messages): Seq[HtmlFormat.Appendable] = {
    statementsForAllEoris.zipWithIndex.flatMap {
      case (statementsForEori, historyIndex) if statementsForEori.currentStatements.nonEmpty =>
        val eoriHeader = generateEoriHeader(historyIndex, statementsForEori)
        val eomHeader = generateEomHeader(statementsForEori)

        val pdfStatements = generateStatements(statementsForEori, historyIndex, isCsv = false)
        val csvStatements = generateStatements(statementsForEori, historyIndex, isCsv = true)

        Seq(eoriHeader, Some(pdfStatements), eomHeader, Some(csvStatements)).flatten

      case _ => Seq.empty
    }
  }

File: test/viewmodels/SecurityStatementsViewModelSpec.scala [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    when(appConfig.historicRequestUrl(SecurityStatement)).thenReturn("request-statement-url")
+    when(appConfig.requestedStatements(SecurityStatement)).thenReturn("request-statement")
+
+    val expectedHeader: HtmlFormat.Appendable =
+      h1Component(msg = "cf.security-statements.title", classes = "govuk-heading-xl")
+
+    val expectedRequestedNotification: HtmlFormat.Appendable =
+      new requestedStatements(linkComponent).apply("request-statement")
+
+    val expectedCurrentStatements: HtmlFormat.Appendable = HtmlFormat.fill(Seq(
```
Comment (tyagikumarvineet) [inline test/viewmodels/SecurityStatementsViewModelSpec.scala:84]: Better to refactor this as it has become very hard to read as it has nested a lot. Better to create dl components outside of this and use it here. If you want to use dl component directly here then create dlComponent's content value outside of this. Will discuss this with you.

File: app/viewmodels/SecurityStatementsViewModel.scala [hunk 6]
```diff
@@ -16,9 +16,259 @@
...
+
+case class SecurityStatementsViewModel(pageTitle: Option[String],
+                                       backLink: Option[String],
+                                       header: HtmlFormat.Appendable,
+                                       requestedStatementNotification: HtmlFormat.Appendable,
+                                       currentStatements: HtmlFormat.Appendable,
+                                       missingGuidance: HtmlFormat.Appendable,
+                                       statementServiceParagraph: HtmlFormat.Appendable,
+                                       requestStatementsLink: HtmlFormat.Appendable,
+                                       details: SecurityStatementsViewModelDetails)
```
Comment (tyagikumarvineet) [inline app/viewmodels/SecurityStatementsViewModel.scala:42]: Could you please advise where this is being used? Seems we are using this only in Spec and in population via SecurityStatementsViewModel's apply method. If it is not being used then we can remove this field.

File: app/viewmodels/SecurityStatementsViewModel.scala [hunk 7]
```diff
@@ -16,9 +16,259 @@
...
+        case _ => Seq.empty
+      })
+    } else {
+      generateNoStatementsParagraph
+    }
+  }
+
+  private def generateEoriHeader(historyIndex: Int, statementsForEori: SecurityStatementsForEori)
+                                (implicit messages: Messages): Option[HtmlFormat.Appendable] = {
+    if (historyIndex > 0) {
```
Comment (AdamCoops) [inline app/viewmodels/SecurityStatementsViewModel.scala:103]: Sometimes we're using if statements and sometimes case statements in these private methods, do we want to just use one if possible and be consistent?

File: app/views/components/description_list/dl.scala.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@this()
+
+@(content: Html, classes: Option[String] = None, id: Option[String] = None)
+
+<dl @id.map(i => s"id=$i") @classes.map(c => s"class=$c")>@content</dl>
```
Comment (AdamCoops) [inline app/views/components/description_list/dl.scala.html:21]: Given these four views are all very similar do we think we could refactor them to be one view where we pass in dl or dt or w/e?

File: app/views/components/span.scala.html [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ *@
+
+@import utils.Utils.emptyString
+
+@this()
+
+@(key: String, classes: Option[String] = None, ariaHidden: Option[String] = None)(implicit messages: Messages)
+
+<span @{classes.fold(emptyString)(classes => s"class=$classes")}
+  @{ariaHidden.fold(emptyString)(value => s"aria-hidden=$value")}>@messages(key)</span>
```
[the remark this answers is not in the document]
Comment (Neozxc) [reply app/views/components/span.scala.html]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: test/views/components/DivSpec.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+import views.html.components.div
+
+class DivSpec extends SpecBase {
+
+  "div component" should {
+
+    "display correct contents" when {
+
+      "only content has been provided" in new Setup {
+        val div: Elements = divComponent.select("div")
```
[the remark this answers is not in the document]
Comment (Neozxc) [reply test/views/components/DivSpec.scala:36]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: test/views/components/H1Spec.scala [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+  "h1 component" should {
+
+    "display correct contents" when {
+
+      "only message key has been provided" in new Setup {
+        h1Component.text() mustBe message(msgKey)
+
+        val h1: Elements = h1Component.select("h1")
+        h1.size() mustBe 1
```
[the remark this answers is not in the document]
Comment (Neozxc) [reply test/views/components/H1Spec.scala:39]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: test/views/components/H1Spec.scala (hunk 2 above)
Comment (Neozxc) [reply test/views/components/H1Spec.scala:64]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: test/views/components/description_list/DescriptionListSpec.scala (hunk 3 above)
Comment (Neozxc) [reply test/views/components/description_list/DescriptionListSpec.scala:36]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: app/viewmodels/SecurityStatementsViewModel.scala (hunk 4 above)
Comment (Neozxc) [reply app/viewmodels/SecurityStatementsViewModel.scala]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: app/viewmodels/SecurityStatementsViewModel.scala (hunk 6 above)
Comment (Neozxc) [reply app/viewmodels/SecurityStatementsViewModel.scala]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: test/viewmodels/SecurityStatementsViewModelSpec.scala (hunk 5 above)
Comment (Neozxc) [reply test/viewmodels/SecurityStatementsViewModelSpec.scala:82]: This has been addressed in the commit [73dfe78](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/commits/73dfe781532fd8bddb0967c9eeb15974440beaf7)

File: app/views/components/description_list/dl.scala.html (hunk 8 above)
Comment (Neozxc) [reply app/views/components/description_list/dl.scala.html:21]: This is similar to Waigwa's question, answered [here](https://github.com/hmrc/customs-financials-documents-frontend/pull/101/files/bf78cc7fead314f8ba2f402316efbe2d9cddf750#diff-c69d8d6cb8f70a75d217dc11661e07a401f1e2ce3518900a312d40b997383b71)

**Review by tyagikumarvineet — APPROVED** [review]

**Review by wg-hmrc — APPROVED** [review]

File: app/viewmodels/SecurityStatementsViewModel.scala (hunk 1 above)
Comment (Neozxc) [reply app/viewmodels/SecurityStatementsViewModel.scala:130]: This has been addressed over chat.

**Pull request merged**